### PR TITLE
Enable Transformer Engine fused cross entropy

### PR DESCRIPTION
**Pull request by eagle705**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

Enable the Transformer Engine fused cross-entropy implementation now that its numerical stability and memory usage issues have been addressed upstream.

This PR:

- Allows `--cross-entropy-loss-fusion --cross-entropy-fusion-impl te` during training argument validation.
- Removes the stale stability warning from `ModelParallelConfig`.
- Updates unit tests and related documentation to treat the TE implementation as a supported configuration.

The previous restriction was introduced because of issues in Transformer Engine's fused cross-entropy implementation. These issues have since been addressed:

- [TransformerEngine#3193](https://github.com/NVIDIA/TransformerEngine/pull/3193) fixes gradient precision loss by preventing FP32 intermediate gradients from being stored in the BF16 input buffer.
- [TransformerEngine#3273](https://github.com/NVIDIA/TransformerEngine/pull/3273) reduces the additional memory introduced by that numerical fix. It saves compact softmax statistics and reconstructs gradients during backward instead of materializing the full gradient during forward.

## Validation

The configuration was validated using a reduced, randomly initialized
Qwen3-30B-A3B-shaped MoE model on 8 H100 GPUs:

- 12 layers, 8.10B total parameters, and 1.30B active parameters
- BF16 training with sequence length 2,048
- TP=2, EP=8, PP=1, and CP=1
- WikiText-103 subset tokenized with the Qwen3 tokenizer
- Transformer Engine built from the merged #3273 commit

Direct native-vs-TE cross-entropy comparisons showed:

- Maximum loss error: `9.5367e-7`
- Maximum logits-gradient error: `1.4901e-8`

Native and TE fused cross entropy were also compared over 30,000 training steps. Both runs completed all 30,000 iterations with zero skipped or NaN iterations. Their losses remained closely correlated, and the mean `TE - native` loss difference over steps 10,001–30,000 was `-0.000246`.

Additional 20-step smoke tests covered TP, PP, EP, and CP configurations.
All 12 native/TE runs completed without skipped or NaN iterations.

These results validate numerical parity, distributed execution, memory usage, and training stability for the tested configuration. This was a toy pretraining experiment with one seed and random initialization, rather than a model-quality evaluation of an official Qwen checkpoint.

image

## Issue tracking

Linked issue: N/A

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] If this PR adds or changes a GPU kernel (Triton, `jit_fuser`/`torch.compile`, CUDA extension, TE or external-library dispatch, or a scatter/index accumulation), I have added or updated its bit-exact determinism test and registered it in `tests/unit_tests/determinism/kernels/manifest.py` ([guide](https://github.com/NVIDIA/Megatron-LM/blob/main/docs/developer/determinism/testing.md))
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [x] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### CODE REVIEW

**Review by Phlip79 — APPROVED** [review]

File: tests/unit_tests/models/test_backend_provider_contracts.py [hunk 1]
```diff
@@ -92,7 +92,7 @@ def test_native_fusion(self):
         assert target is fused_vocab_parallel_cross_entropy
 
     def test_an_explicit_te_request_is_not_silently_ignored(self):
```
Comment (deepakn94) [inline tests/unit_tests/models/test_backend_provider_contracts.py]: Nit: remove "test_an_explicit" -> "test_explicit"

Comment (eagle705) [reply tests/unit_tests/models/test_backend_provider_contracts.py]: @deepakn94 @Phlip79 Done, renamed `test_an_explicit` to `test_explicit`.

File: megatron/training/arguments.py
Comment (jaredcasper) [inline megatron/training/arguments.py:1788]: Can we add a TE version check here?
